### DISCUSSION TITLE
2024-07-30 화요일

### DISCUSSION THREAD

**Discussion by eunohhh** [nbc-final-a10/TripBuddy#141] [기록보드]
2024-07-29 화요일 기록보드입니다.

## 여행가고십조(A10) 일일 공통 일과표

| 시간 | 일과 |  기록 |
| :-: | :-: | :-: |
| 09:00 ~ 10:00 | 아침 회의 | |
| 10:00 ~ 11:00 | 기술면접 준비 | |
| 11:00 ~ 13:00 | 집중 개발 time | |
| 13:00 ~ 14:00 | 점심 식사 | |
| 14:00 ~ 18:00 | 집중 개발 time | |
| 18:00 ~ 19:00 | 저녁 식사 | |
| 19:00 ~ 20:00 | 저녁 회의 | |
| 20:00 ~ 21:00 | 5분 기록보드 작성 / 개인 회고 | |

```
## 개인 기록보드 서식 (김병준)

### 오늘 한 일

- 내용

### 트러블 슈팅

- 내용

### 내일 계획

- 내용
```

```
// 이미지 테이블
<table>
  <tr>
    <th>이미지 1번 제목</th>
    <th>이미지 2번 제목</th>
  </tr>
  <tr>
    <td><img src="이미지 주소" width="300" height="300"/></td>
    <td><img src="이미지 주소" width="300" height="300"/></td>
  </tr>
</table>
```

**Comment by Captain-Kim** [comment]
## 개인 기록보드 서식 (김병준)

### 오늘 한 일

- 글쓰기 페이지 기능 구현, API Route 작성
- 글쓰기 페이지 추가 검색 옵션 컴포넌트 제작

### 트러블 슈팅

- px 단위가 clsx 설정에서 동적으로 할당되지 않음 -> 수치가 먹히긴 하나 유틸리티 클래스가 먹히지 않음 -> vh는 됨 (스크롤 생성 목적의 동적 스타일링)
- 커스텀 훅 작성 패턴 문제
  - 커스텀 훅에서 컴포넌트를 정의하고 해당 커스텀 훅과 컴포넌트를 return 하는 패턴으로 내내 커스텀 훅을 작성하였는데
  - 공유용은 아니고 글쓰기 페이지의 state 값이 필요하여 혼자 사용할 커스텀 훅으로 변경하여 제작하였더니
  - return한 컴포넌트 내부의 입력필드 (제목, 글내용)에서 한 글자를 칠 때마다 포커스를 잃는 기괴한 현상 발생
  - **문제의 원인은** 커스텀 훅 내부에서 컴포넌트를 정의하였기 때문. 훅 내부에 컴포넌트를 정의하면 컴포넌트가 리렌더링 될 때마다 컴포넌트의 인스턴스가 새롭게 생성됨. 이것 때문에 한 글자를 칠 때마다 -> setState에 의해 컴포넌트 리렌더링이 발생하고 -> 컴포넌트의 인스턴스가 새롭게 생성되면서 -> 이 컴포넌트의 모든 DOM 요소들이 새롭게 생성됨 -> 입력필드 또한 DOM 요소이므로 포커스를 잃게 됨
  - **해결한 방법은** 커스텀 훅을 정의할 때 컴포넌트를 훅 외부에서 별도로 정의해서 컴포넌트가 재정의되는 것을 방지함.
    - 훅에서는, state와 handler만 return. 컴포넌트는 별도로 작성
    - 컴포넌트에서는, 커스텀 훅을 호출해서 state와 handler를 props로 받아서 사용
  - 시도해보진 않았지만 컴포넌트에서 state와 handler를 다 정의하면 편하지 않느냐? -> 자식 컴포넌트에서 부모 컴포넌트로 props를 올리는 행위는 패륜이기 때문에 시도해보지 않음
- 이 글 다 쓰고 든 생각인데 **비제어컴포넌트로 만들었으면** out of focus 문제가 생기지 않았을 수도 있겠다라는 생각이 드네...

```tsx
// 커스텀 훅
'use client';

import React, { useState } from 'react';

function useTripWrite() {
    const [tripTitle, setTripTitle] = useState('');
    const [tripContent, setTripContent] = useState('');
    const [tripImage, setTripImage] = useState('');

    const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTripTitle(e.target.value);
    };
    const handleContentChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        setTripContent(e.target.value);
    };

    return {
        tripTitle,
        tripContent,
        tripImage,
        handleTitleChange,
        handleContentChange,
    };
}

export default useTripWrite;
```
```tsx
// 컴포넌트
'use client';

import React from 'react';
import Left2xlBoldText from '@/components/atoms/write/Left2xlText';

const WriteTrip: React.FC<{
    tripTitle: string;
    tripContent: string;
    handleTitleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
    handleContentChange: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
}> = ({ tripTitle, tripContent, handleTitleChange, handleContentChange }) => {
    return (
        <div className="p-4">
            <header className="mb-5">
                <Left2xlBoldText text="모집 글을 작성해봐요!" />
            </header>

            <form className="space-y-4">
                <div>
                    <label className="block mb-1 text-sm font-medium text-gray-700">
                        대표 이미지
                    </label>
                    <button className="flex items-center justify-center w-20 h-20 bg-gray-200 border border-gray-300 rounded">
                        <span className="text-gray-400">📷</span>
                    </button>
                </div>
                <div>
                    <label className="block mb-1 text-sm font-medium text-gray-700">
                        제목
                    </label>
                    <input
                        type="text"
                        value={tripTitle}
                        onChange={handleTitleChange}
                        placeholder="제목"
                        maxLength={20}
                        className="w-full px-3 py-2 border border-gray-300 rounded"
                    />
                    <span className="block text-right text-sm text-gray-500">{`${tripTitle.length}/20`}</span>
                </div>
                <div>
                    <label className="block mb-1 text-sm font-medium text-gray-700">
                        글 내용
                    </label>
                    <textarea
                        value={tripContent}
                        onChange={handleContentChange}
                        placeholder="내용을 입력해주세요."
                        className="w-full h-96 px-3 py-2 border border-gray-300 rounded resize-none"
                    />
                </div>
            </form>
        </div>
    );
};

export default WriteTrip;
```
```tsx
// 페이지 컴포넌트에서 컴포넌트를 호출할 때
// 커스텀 훅을 먼저 호출해서 state 등을 꺼내온 후
// 호출한 state, 함수는 전부 컴포넌트에 Props로 내려주어야 함
<WriteTrip
    tripTitle={tripTitle}
    tripContent={tripContent}
    handleTitleChange={handleTitleChange}
    handleContentChange={handleContentChange}
/>
```

### 궁금한 점

<img src="https://github.com/user-attachments/assets/f057149e-8aa4-4491-95d9-b19f81951fbf" height="300" width="300"/>

- 페이지 컴포넌트 : 서버 컴포넌트
- 자식 컴포넌트 : 모두 클라이언트 컴포넌트

=> 이렇게 구성해도 SEO 최적화에서 Metadata만 잘 설정하면 괜찮은가?

### 내일 계획

- 마이페이지 UI 완성
- 마이페이지 기능 구현

**Reply by eunohhh** [reply]
오호! 이점을 생각하지 못했군요!
그렇다면 혹시 React.memo 처리를 하면 문제가 해결 될 수도(??)있겠네요

**Reply by Captain-Kim** [reply]
> 오호! 이점을 생각하지 못했군요! 그렇다면 혹시 React.memo 처리를 하면 문제가 해결 될 수도(??)있겠네요

오 최적화 훅은 사용해본 적이 없는데 UT 이후 리팩토링 하면서 한 번 시도 해보겠습니다 감사합나디~~!!

**Comment by eunohhh** [comment]
## 개인 기록보드 서식 (오은)

### 오늘 한 일

- onboarding 페이지 기능 완성

### 트러블 슈팅

#### 온보딩 관련
- 온보딩은 isOnboarding이 false일 때 한번만 되는 것인데 마지막 step 에서 한번에 fetch 하게 되면 
- 모종의 이유로 에러가 발생해서 시퀀스가 중단 되었을 경우에 아무 정보도 fetch 하지 못한 상태가 되기 때문에 
- 각 step 마다 fetch 하는 것으로 결정

#### 온보딩 관련
- 네이버 스크립트 루트레이아웃에서 너무 많은 리소스 차지
- 하나씩 레벨 내려보면서 테스트할것

### 내일 계획

- 스토리 관련 진행

**Reply by Captain-Kim** [reply]
**<<트러블슈팅>>** 오 그러면 마지막 시퀀스에 연결된 handler에서 isOnboarding을 토글하게 되나요?

**Reply by eunohhh** [reply]
네 맞습니다~!

**Comment by Captain-Kim** [comment]
# 추가기입

## 개인 기록보드 서식 (김병준)

### 오늘 한 일

- Funnel Patter 마지막 시퀀스로 state 넘기도록 전면 리팩토링
- Supabase 데이터 테이블 재점검 (외래 키 설정 오류 발견)

### 트러블 슈팅

- 확실하지 않다면 SQL로 일괄적으로 데이터 테이블을 생성할 때 반드시 CRUD가 잘 되는지 확인해봐야 한다.
- Funnel Pattern 에서 각 시퀀스에서 사용자에게 입력받은 state를 어떻게 마지막 시퀀스로 전달할 것인가?
  - -> children 간 props 전달 불가
  - -> 전역 상태 관리를 해야 하나? -> 검색 페이지에서도 재사용되는 커스텀 훅이기 때문에 중복되는 state가 발생할 것 같음. -> 만약 이렇게 되면 검색 페이지에서 지역 검색을 '일본'으로 선택한다면 글쓰기 페이지에 와서도 그 전역 상태가 공유 되어서 사용자가 값을 입력하지 않았음에도 initialState가 '일본'이 되어버리는 현상이 발생할 것 같음.
  - -> 그러면 검색 페이지와 글쓰기 페이지에서 store를 다르게 하면 되지 않을까? -> 가능할 것 같지만 Funnel Pattern도 컴포넌트의 흐름을 파악하기 힘들 정도로 컴포넌트가 쏟아지는데 store까지 왔다 갔다 하다 보면 오히려 흐름이 더 복잡해질 것 같음.
  - -> Funnel Pattern에 시퀀스로 호출된 자식 컴포넌트에서 커스텀 훅을 호출하는 것이 아니라 페이지 컴포넌트 단에서 커스텀 훅을 호출한 뒤 글 제출이 이루어지는 마지막 시퀀스에 Props로 전달하면 되지 않을까? -> 잘 됨. 오히려 이 방식이 페이지 컴포넌트에서 단방향 데이터 흐름을 캐치하기 편한 것 같음.
  - -> 그러면 전역 상태 관리를 하지 않고 Props로 관리하는 경우 단점은 무엇인가? -> 모르겠음. Props Drilling이 발생하는 경우가 아니면 메모리 관리 관점에서도 굳이 글쓰기 페이지와 검색 페이지의 store를 이름만 다르고 같은 로직으로 만드는 각각 생성해야 하는 전역 상태 관리보다 단순 Props를 전달하는 것이 더 나은 선택인 것 같음.
  - -> 그런데 마지막 시퀀스에 너무 많은 Props가 전달 되는 것 같다 -> 페이지 컴포넌트 단에서 핸들러 함수를 정의해서 supabase에 POST 요청을 보내는 함수를 묶어서 작성하고 그 함수 하나만 마지막 시퀀스에 내린다 -> 마지막 시퀀스에서 '글 작성하기' 버튼 등에 이벤트 핸들러를 연결해서 제출만 한다.
- 더 알아볼 점 : Props는 단방향 데이터 통신일 텐데 페이지 컴포넌트(부모) - 자식 컴포넌트(시퀀스) 간에 어떻게 state의 값이 계속 주고받을 수 있지?

<table>
  <tr>
    <th>커스텀 훅 페이지 컴포넌트에서 호출하는 방식으로 리펙토링 중</th>
    <th>진짜 잘 되는 모습</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/38c2f40c-f242-41b5-9812-bda6ffad6b83" width="300" height="300"/></td>
    <td><img src="https://github.com/user-attachments/assets/c3ae676c-f363-4402-a70e-e276f0a09564" width="300" height="300"/></td>
  </tr>
</table>

### 내일 계획

- 내용

**Comment by hyunjeongkwak** [comment]
**오늘 한 일**
- 탑 버튼, 자동 스크롤 기능
- 성별 필터 기능 진행 중('검색 결과 보기' 버튼을 눌렀을 때 렌더링 되게 수정해야 함. 현재 칩을 누른 즉시 변경됨)

**내일 할 일**
- 성별 필터 기능
- 만남 장소, 인원수 필터 기능
- 가능한 필터 기능까지 (남은 것: 인원수, 만남 장소, 여행지, 날짜, 여정 테마, 버디즈 성향, 나이, 타이핑 쳐서 검색)

**Comment by slsl2** [comment]
## 개인 기록보드 (김민지)

### 오늘 한 일

- 채팅 realtime 구현
- 채팅 페이지 스타일 작업

### 미해결 이슈

- 프로필 사진이 새로고침 하기 전에 안 뜨는 문제

### 내일 할 일

- 이슈 해결
- chat 목록 페이지 데이터 연결
- 채팅 페이지 헤더 데이터 연결
- 채팅 기능 디테일 잡기
